File: texture-generator/pre/colorfilter/MetalColorFilter.ts
class MetalColorFilter extends ColorFilter {

    private color: number;

    constructor(color: string, private shinyness: number, private brightness: number, private hueshift: number){
        super();
        this.color = Color.parseColor(color);
    }

    preProcess(data: number[]): void {

    }

    colorPixel(pixel: number, index: number): number {
        const a = ColorFilter.getA(pixel);
        if(a === 0){
            return pixel;
        }
        const l = ColorFilter.getPerceptualBrightness(pixel) / 255;
        const ll = l ** 2;
        const r = ColorFilter.mult(ColorFilter.getR(this.color), ColorFilter.getR(pixel)) & 0xff;
        const g = ColorFilter.mult(ColorFilter.getG(this.color), ColorFilter.getG(pixel)) & 0xff;
        const b = ColorFilter.mult(ColorFilter.getB(this.color), ColorFilter.getB(pixel)) & 0xff;
        const hsb = ColorFilter.colorToHSB(r, g, b);
        hsb[0] -= (0.5 - ll) * this.hueshift;
        while(hsb[0] < 0){
            hsb[0] += 360;
        }
        while(hsb[0] >= 360){
            hsb[0] -= 360;
        }
        if(l > 0.9) {
            hsb[1] = Math.min(Math.max(hsb[1] - ll * this.shinyness, 0), 1);
        }
        if(l > 0.8) {
            hsb[2] = Math.min(Math.max(hsb[2] + ll * this.brightness, 0), 1);
        }
        return Color.HSVToColor(a, hsb);
    }

    postProcess(data: number[]): void {

    }

}

File: texture-generator/filters/metal_color_filter.py
import colorsys
from PIL import ImageColor
from . import BaseFilter, RGBA
import utils


class MetalColorFilter(BaseFilter):


    def __init__(self, color_hex: str, shinyness: float, brightness: float, hueshift: float):
        self.color = ImageColor.getcolor(color_hex, "RGBA")
        self.shinyness = shinyness
        self.brightness = brightness
        self.hueshift = hueshift


    def color_pixel(self, rgba: RGBA, index: int) -> RGBA:

        if rgba[3] == 0:
            return rgba
        
        l = utils.perceptual_brightness(rgba) / 255
        ll = l ** 2
        r = utils.mult_color(self.color[0], rgba[0]) & 0xff
        g = utils.mult_color(self.color[1], rgba[1]) & 0xff
        b = utils.mult_color(self.color[2], rgba[2]) & 0xff
        h, s, v = colorsys.rgb_to_hsv(r / 255, g / 255, b / 255)

        h -= (0.5 - ll) * self.hueshift
        while h < 0: h += 1.0
        while h >= 1.0: h -= 1.0

        if l > 0.9: s = min(max(s - ll * self.shinyness, 0.0), 1.0) # 0.0 - 1.0
        if l > 0.8: v = min(max(v + ll * self.brightness, 0.0), 1.0) # 0.0 - 1.0

        r, g, b = colorsys.hsv_to_rgb(h, s, v)

        return (int(r * 255), int(g * 255), int(b * 255), rgba[3])
